### PR TITLE
DEVX-155 migrate state sync

### PR DESCRIPTION
**Pull request by lojzatran**
#### Summary

JIRA: https://commercetools.atlassian.net/browse/DEVX-155

### CODE REVIEW

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 1]
```diff
@@ -0,0 +1,1285 @@
...
+            .get()
+            .execute()
+            .thenApply(ApiHttpResponse::getBody)
+            .join();
+
+    assertThat(fetchedState.getRoles()).isEmpty();
+  }
+
+  @Test
+  void sync_withNewStateWithoutRole_shouldDoNothing() {
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:184]: the name is missleading, the test above is named `sync_withNewStateWithoutRole_shouldRemoveRole`, which describes the same setup with different outcome. Please adjust.

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 2]
```diff
@@ -0,0 +1,1285 @@
...
+  void sync_withEqualState_shouldNotUpdateState() {
+    StateDraft stateDraft =
+        StateDraftBuilder.of()
+            .key(key)
+            .type(StateTypeEnum.LINE_ITEM_STATE)
+            .name(ofEnglish("state-name"))
+            .description(ofEnglish("state-desc"))
+            .roles(StateRoleEnum.RETURN)
+            .initial(false)
+            .build();
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:317]:
```suggestion
String key = this.key;
    StateDraft stateDraft =
        StateDraftBuilder.of()
            .key(key)
            .type(StateTypeEnum.LINE_ITEM_STATE)
            .name(ofEnglish("state-name"))
            .description(ofEnglish("state-desc"))
            .roles(StateRoleEnum.RETURN)
            .initial(false)
            .build();
```

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 3]
```diff
@@ -0,0 +1,1285 @@
...
+                      format(
+                          "Failed to resolve 'transition' reference on StateDraft with key:'%s",
+                          keyC));
+              return null;
+            })
+        .toCompletableFuture()
+        .join();
+  }
+
+  private State createStateInSource(final StateDraft draft) {
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:1047]: Maybe you can split the tests using a source - ctp - project and put to /internalsource/states ?

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 4]
```diff
@@ -0,0 +1,1285 @@
...
+    final StateDraft stateDraft =
+        StateDraftBuilder.of().key(keyA).type(StateTypeEnum.REVIEW_STATE).roles(List.of()).build();
+
+    final StateDraft stateDraftTarget =
+        StateDraftBuilder.of()
+            .key(keyA)
+            .type(StateTypeEnum.REVIEW_STATE)
+            .roles(List.of(StateRoleEnum.REVIEW_INCLUDED_IN_STATISTICS))
+            .build();
+    createStateInTarget(stateDraftTarget);
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:159]: why do you need another state in target with `keyA` ?

Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:159]: why do you need another state in target with `keyA` ?

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 5]
```diff
@@ -0,0 +1,1285 @@
...
+    final StateDraft stateDraft =
+        StateDraftBuilder.of()
+            .key(keyA)
+            .type(StateTypeEnum.REVIEW_STATE)
+            .roles(List.of(StateRoleEnum.REVIEW_INCLUDED_IN_STATISTICS))
+            .build();
+
+    final StateDraft stateDraftTarget =
+        StateDraftBuilder.of().key(keyA).type(StateTypeEnum.REVIEW_STATE).roles(List.of()).build();
+    createStateInTarget(stateDraftTarget);
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:217]: why do you need another state in target with keyA ?

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 6]
```diff
@@ -0,0 +1,1285 @@
...
+  static void tearDown() {
+    deleteStatesFromTargetAndSource();
+  }
+
+  @BeforeEach
+  void setup() {
+    key = "state-" + current().nextInt();
+    keyA = "state-A-" + current().nextInt();
+    keyB = "state-B-" + current().nextInt();
+    keyC = "state-C-" + current().nextInt();
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:76]: Not sure why those keys are needed global, but not locally?

File: src/main/java/com/commercetools/sync/sdk2/services/impl/BaseService.java [hunk 7]
```diff
@@ -60,7 +60,6 @@ abstract class BaseService<
 
   protected static final int MAXIMUM_ALLOWED_UPDATE_ACTIONS = 500;
   static final String CREATE_FAILED = "Failed to create draft with key: '%s'. Reason: %s";
-  static final String FETCH_FAILED = "Failed to fetch resource with key: '%s'. Reason: %s";
```
Comment (salander85) [inline src/main/java/com/commercetools/sync/sdk2/services/impl/BaseService.java:63]: 😄

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+import com.commercetools.api.models.state.StateDraft;
+import java.util.HashSet;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+
+class WaitingToBeResolvedTransitionsTest {
+
+  @Test
+  void setProductDraft_WithNonNullProductDraft_ShouldSetProductDraft() {
```
Comment (salander85) [inline src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java:16]:
```suggestion
  void setStateDraft_WithNonNullStateDraft_ShouldSetStateDraft() {
```

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+    // test
+    waitingToBeResolvedTransition.setStateDraft(stateDraft);
+
+    // assertions
+    assertThat(waitingToBeResolvedTransition.getStateDraft()).isEqualTo(stateDraft);
+  }
+
+  @Test
+  void setMissingReferencedProductKeys_WithNonNullSet_ShouldSetTheSet() {
```
Comment (salander85) [inline src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java:30]:
```suggestion
  void setMissingTransistionStateKeys_WithNonNullSet_ShouldSetTheSet() {
```

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+    // test
+    boolean result = waitingToBeResolvedTransition.equals(other);
+
+    // assertions
+    assertFalse(result);
+  }
+
+  @Test
+  void equals_WithDifferentProductDraft_ShouldReturnFalse() {
```
Comment (salander85) [inline src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java:87]:
```suggestion
  void equals_WithDifferentStateDraft_ShouldReturnFalse() {
```

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    // test
+    final int hash1 = waitingToBeResolvedTransition.hashCode();
+    final int hash2 = other.hashCode();
+
+    // assertions
+    assertNotEquals(hash1, hash2);
+  }
+
+  @Test
+  void hashCode_withSameProductKeyAndDiffRefSet_ShouldNotBeEquals() {
```
Comment (salander85) [inline src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java:175]:
```suggestion
  void hashCode_withSameStateKeyAndDiffTransistionSet_ShouldNotBeEquals() {
```

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    // test
+    final int hash1 = waitingToBeResolvedTransition.hashCode();
+    final int hash2 = other.hashCode();
+
+    // assertions
+    assertEquals(hash1, hash2);
+  }
+
+  @Test
+  void hashCode_withDifferentProductKeyAndSameRefSet_ShouldNotBeEquals() {
```
Comment (salander85) [inline src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java:155]:
```suggestion
  void hashCode_withDifferentStateKeyAndSameTransitionSet_ShouldNotBeEquals() {
```

**Review by salander85 — CHANGES_REQUESTED** [review]

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java [hunk 13]
```diff
@@ -0,0 +1,1285 @@
...
+
+  private List<String> errorCallBackMessages;
+  private List<String> warningCallBackMessages;
+  private List<Throwable> errorCallBackExceptions;
+  private String key = "";
+  private ReferenceIdToKeyCache referenceIdToKeyCache;
+
+  @AfterAll
+  static void tearDown() {
+    deleteStatesFromTargetAndSource();
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/19d7c3fd5dcc5e0058a8e4e9860092feb5b0b248

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java (hunk 1 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/19d7c3fd5dcc5e0058a8e4e9860092feb5b0b248

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java (hunk 2 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/d290cc2fa9e10087084d1030192ba8b2c6a6b913

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java (hunk 3 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/cdcd860d7e04c223366495161bfffe5de85fb8e4

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java (hunk 4 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java:146]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/542ff2af53cf56aaccdd8cd76b645c848672abe3

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java (hunk 9 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/de5342932f65dabdc0391e75607e1558d7472990

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java (hunk 10 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/de5342932f65dabdc0391e75607e1558d7472990

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java (hunk 11 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/de5342932f65dabdc0391e75607e1558d7472990

File: src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java (hunk 12 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/sdk2/commons/models/WaitingToBeResolvedTransitionsTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/de5342932f65dabdc0391e75607e1558d7472990

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/states/StateSyncIT.java [hunk 14]
```diff
@@ -0,0 +1,666 @@
...
+  private String keyC;
+
+  private List<String> errorCallBackMessages;
+  private List<String> warningCallBackMessages;
+  private List<Throwable> errorCallBackExceptions;
+  private ReferenceIdToKeyCache referenceIdToKeyCache;
+
+  @AfterAll
+  static void tearDown() {
+    deleteStatesFromTarget();
```
Comment (salander85) [inline src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/states/StateSyncIT.java:60]: here you need to cleanup source too :)

**Review by salander85 — APPROVED** [review]
Very nice!

File: src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/states/StateSyncIT.java (hunk 14 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/states/StateSyncIT.java]: https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/85e629e0636557ef91a6979bd713881a7d798e77

**Pull request merged**